DOC Form sudo mode (#684)
silverstripe/developer-docs: en/02_Developer_Guides/09_Security/04_Sudo_Mode.md

Sudo mode is not active when a user logs in to the CMS. If the PHP session lifetime expires before the sudo mode lifetime, that sudo mode will also be cleared. If the user leaves their CMS open, or continues to use it, for an extended period of time with automatic refreshing in the background, sudo mode will eventually deactivate once the max lifetime is reached.

Data protected by sudo mode is still viewable without entering the password, but any actions that modify data will require the user to enter their password. This is done to balance usability with security. The following [`DataObject`](api:SilverStripe\ORM\DataObject) subclasses are protected by sudo mode:

- [`Member`](api:SilverStripe\Security\Member)
- [`Group`](api:SilverStripe\Security\Group)
- [`PermissionRole`](api:SilverStripe\Security\PermissionRole)
- [`PermissionRoleCode`](api:SilverStripe\Security\PermissionRoleCode)
- [`SiteConfig`](api:SilverStripe\SiteConfig\SiteConfig)

## Adding or removing sudo mode for a `DataObject`

> To add sudo mode for a particular `DataObject`, simply set the [`DataObject.require_sudo_mode`](api:SilverStripe\ORM\DataObject->require_sudo_mode) configuration property to `true`, either directly on the class or via yml.

Reviewer: ```suggestion
To add sudo mode for a particular model, including your `DataObject` subclasses, simply set the [`require_sudo_mode`](api:SilverStripe\View\ViewableData->require_sudo_mode) configuration property to `true`, either directly on the class or via yml.
```

Hopefully this strikes a good balance between being accurate (the API doc link should be for where the config is defined) and not confusing (we don't actually mention `ViewableData` in the copy directly)

Author: Updated

```php
namespace App\Model;

use SilverStripe\ORM\DataObject;

class Product extends DataObject
{
    // ...
    private static bool $require_sudo_mode = true;
}
```

```yml
SomeModule\Model\Player:
  require_sudo_mode: true
```

> [!NOTE]

> > This will only add sudo mode to `DataObject` edit forms within the CMS interface. It will have no effect on forms outside of the CMS, such as custom forms in the frontend.

Reviewer: ```suggestion
> This will only add sudo mode to edit forms within the CMS interface. It will have no effect on forms outside of the CMS, such as custom forms in the frontend.
```

Author: Updated

# Sudo mode

Sudo mode represents a heightened level of permission in that you are more certain that the current CMS user is actually the person whose account is logged in. This is performed by re-validating that the account's password is correct, and will then last for a certain amount of time (configurable) until it will be checked again.

Sudo mode is not active when a user logs in to the CMS. If the PHP session lifetime expires before the sudo mode lifetime, that sudo mode will also be cleared. If the user leaves their CMS open, or continues to use it, for an extended period of time with automatic refreshing in the background, sudo mode will eventually deactivate once the max lifetime is reached.

## Sudo mode for models

> The data of models (usually a [`DataObject`](api:SilverStripe\ORM\DataObject) subclass) are protected by sudo mode is still viewable without entering the password, and any actions that modify data, e.g. POSTing an edit form submission, will require the user to enter their password. This is done to balance usability with security. The following `DataObject` subclasses are protected by sudo mode:

Reviewer: ```suggestion
Modules which are protected by sudo mode (usually a [`DataObject`](api:SilverStripe\ORM\DataObject) subclass) are still viewable without entering the password. Any actions that modify data, e.g. POSTing an edit form submission, will require the user to enter their password. This is done to balance usability with security.

The following `DataObject` subclasses are protected by sudo mode out of the box:
```
Just rewording/rearranging this to be a bit clearer

Author: Done, the first word "Modules" I assumed you meant "Models"

Reviewer: Yes, sorry. Thanks for catching that.

- [`Member`](api:SilverStripe\Security\Member)
- [`Group`](api:SilverStripe\Security\Group)
- [`PermissionRole`](api:SilverStripe\Security\PermissionRole)
- [`PermissionRoleCode`](api:SilverStripe\Security\PermissionRoleCode)

> ### Configuring sudo mode for a `DataObject`

Reviewer: ```suggestion
### Configuring sudo mode for your models
```

Author: Updated

To add sudo mode for a particular model, including your `DataObject` subclasses, simply set the [`require_sudo_mode`](api:SilverStripe\View\ViewableData->require_sudo_mode) configuration property to `true`, either directly on the class or via yml.

```php
namespace App\Model;

use SilverStripe\ORM\DataObject;

class Product extends DataObject
{
    // ...
    private static bool $require_sudo_mode = true;
}
```

```yml
SomeModule\Model\Player:
  require_sudo_mode: true
```

> [!NOTE]
> This will only add sudo mode to edit forms within the CMS interface. It will have no effect on forms outside of the CMS, such as custom forms in the frontend.

> > [!WARNING] If you add new fields to form that is protected by sudo mode, such as in an overridden [`LeftAndMain::getEditForm()`](api:SilverStripe\Admin\LeftAndMain::getEditForm()) method, it may mean having to call [`Form::requireSudoMode()`](api:SilverStripe\Forms\Form::requireSudoMode()) on the form to ensure the newly added fields are set to read-only while sudo mode is inactive.

Reviewer: IIRC our docs site doesn't like two callout blocks being right after one another. Consider putting one of these above the code blocks.

Author: Updated

## Configuring the sudo mode lifetime

The default [`SudoModeServiceInterface`](api:SilverStripe\Security\SudoMode\SudoModeServiceInterface) implementation is [`SudoModeService`](api:SilverStripe\Security\SudoMode\SudoModeService), and its lifetime can be configured with YAML. You should read the lifetime value using `SudoModeServiceInterface::getLifetime()`.

```yml
SilverStripe\Security\SudoMode\SudoModeService:
  lifetime_minutes: 25

> ```

Reviewer: This section shouldn't be splitting up the various "sudo mode for" sections. Please move it to be either before or after that group of sections.

Author: Have moved it above the other sections

# Sudo mode

Sudo mode represents a heightened level of permission in that you are more certain that the current CMS user is actually the person whose account is logged in. This is performed by re-validating that the account's password is correct, and will then last for a certain amount of time (configurable) until it will be checked again.

Sudo mode is not active when a user logs in to the CMS. If the PHP session lifetime expires before the sudo mode lifetime, that sudo mode will also be cleared. If the user leaves their CMS open, or continues to use it, for an extended period of time with automatic refreshing in the background, sudo mode will eventually deactivate once the max lifetime is reached.

> Data protected by sudo mode is still viewable without entering the password, but any actions that modify data will require the user to enter their password. This is done to balance usability with security. The following [`DataObject`](api:SilverStripe\ORM\DataObject) subclasses are protected by sudo mode:

Reviewer: > Data protected by sudo mode is still viewable without entering the password

Is that true for the per-field sudo mode as well? I think it isn't, right? The HOC just replaces the field outright?

We need to remember (and point out) that there's now two ways to use sudo mode - on a whole data object, or just for a single form field - and they work subtly differently.

Might make sense to just outright separate them into two separate headings:

1. ``## Protecting an entire model``
1. ``## Protecting a single `FormField` ``

Author: The HOC is basically irrelevant at this point as it only used for the MFA field on the Member form, though the entire Member form is now protected.  You need to enable sudo mode on the form to see the HOC ... which won't be used because sudo mode is now activated

There may be some edge case where a project/module somehow uses the HOC for something else ... though my assumption is that there are zero instances of that so mentioning that here is just adding confusion for no benefit

Reviewer: The HOC is not irrelevant. There are plenty of scenarios where a project or module might want to protect a single sensitive form field with sudo mode. Just because we're not using it directly anymore doesn't make it useless. It's also _still there_ so should be documented regardless of our feelings about it until/if we decide to remove it, which would be a separate card.

> mentioning that here is just adding confusion for no benefit

I also don't see how there would be any confusion provided you use separate sections like I suggested above for the two use cases.

Author: There's currently a "Using sudo mode in a react component" section lower down in this file, which is instructions to use the HOC, which seems like the "protecting a single FormField", scenario, but it's not really.  The MFA field is really a separate xhr field that not part of the form, i.e there's no underlying `<input>` that submitted as part of main form submission.  There's currently no way to have sudo mode on a single regular form field.

I don't think there needs to be any changes made here

Reviewer: There's still a question outstanding which was the start of this whole thread:

>> Data protected by sudo mode is still viewable without entering the password
>
> Is that true for the per-field sudo mode as well? I think it isn't, right? The HOC just replaces the field outright?

If we're not splitting this up into clear sections that make it obvious when we're talking about the per-field implementation vs the entire form implementation, then we need to make it clear in the writing what applies to what. "sudo mode" is the name for both implementations, so we can't just blanket say "sudo mode won't hide data"

Author: I've mostly reorganized the existing headings to make the split clearer

- [`Member`](api:SilverStripe\Security\Member)
- [`Group`](api:SilverStripe\Security\Group)
- [`PermissionRole`](api:SilverStripe\Security\PermissionRole)
- [`PermissionRoleCode`](api:SilverStripe\Security\PermissionRoleCode)
- [`SiteConfig`](api:SilverStripe\SiteConfig\SiteConfig)

## Adding or removing sudo mode for a `DataObject`

To add sudo mode for a particular `DataObject`, simply set the [`DataObject.require_sudo_mode`](api:SilverStripe\ORM\DataObject->require_sudo_mode) configuration property to `true`, either via yml or directly on the class.

While it's not recommended, to remove sudo mode for a particular `DataObject`, set the [`DataObject.require_sudo_mode`](api:SilverStripe\ORM\DataObject->require_sudo_mode) configuration property to `false` via yml.

```yml
SilverStripe\Security\Member:
  require_sudo_mode: false

> ```

Reviewer: ````suggestion
To add sudo mode for a particular `DataObject`, simply set the [`DataObject.require_sudo_mode`](api:SilverStripe\ORM\DataObject->require_sudo_mode) configuration property to `true`, either via yml or directly on the class.

```php
namespace App\Model;
use SilverStripe\ORM\DataObject;
class Product extends DataObject
{
    // ...
    private static bool $require_sudo_mode = true;
}
```

```yml
SomeModule\Model\Player:
  require_sudo_mode: true
```
````
I don't think we should tell people they can disable this and put that idea into their heads. Saying how to configure it in the broader sense (i.e. this is controlled by `DataObject.require_sudo_mode` which you can configure) should be enough to tell people how to turn it off without telling them "you can turn it off"

Author: Updated

  lifetime_minutes: 25
```

> ## Enabling sudo mode for controller endpoints

Reviewer: Just as a sanity check in case you didn't already - please quickly check if there are any links in the docs to the old anchor, and update them to the new anchor if there are.

Author: Checked and did not find any

## Enabling sudo mode for controller endpoints

You can add the `SudoModeServiceInterface` singleton as a dependency to a controller that requires sudo mode for one of its actions:
Performing a sudo mode verification check in a controller endpoint by using the sudo mode service to validate the request:

```php
namespace App\Control;

class MyController extends Controller
{
    private ?SudoModeServiceInterface $sudoModeService = null;

    private static array $dependencies = ['SudoModeService' => '%$' . SudoModeServiceInterface::class];

    public function setSudoModeService(SudoModeServiceInterface $sudoModeService): static
    {
        $this->sudoModeService = $sudoModeService;
        return $this;
    }
}

> ```

Reviewer: Why is this example being removed?

Author: I really dislike the `$dependencies` way of doing it. It's long and unintuitive. I updated the example below to:

```php
$service = Injector::inst()->get(SudoModeServiceInterface::class);
if (!$service->check($request->getSession())) {
// ...
```

Reviewer: I personally think we should keep it, and that this change is unrelated - but in the interest of moving forward I'll let this go
